File: 236.lowest-common-ancestor-of-a-binary-tree.go
/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
 func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    ancestor := make(chan *TreeNode, 1)
    go findAncestor(root, p, q, ancestor)
    return <- ancestor
}

func findAncestor(root, p, q *TreeNode, ancestor chan *TreeNode) {
    if root == nil {
        return
    }
    findAncestor(root.Left, p, q, ancestor)
    findAncestor(root.Right, p, q, ancestor)
    if isAncestor(root.Left, p) && isAncestor(root.Left, q) {
        ancestor <- root.Left
    }
    if isAncestor(root.Right, p) && isAncestor(root.Right, q) {
        ancestor <- root.Right
    }
    if isAncestor(root, p) && isAncestor(root, q) {
        ancestor <- root
    }
}

func isAncestor(root, p *TreeNode) bool {
    if root == nil {
        return false
    }
    if root.Val == p.Val {
        return true
    }
    return isAncestor(root.Left, p) || isAncestor(root.Right, p)
}

File: 236.lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    static class Triple {
        TreeNode root;
        boolean p;
        boolean q;

        public Triple() {
            this.root = null;
            this.p = false;
            this.q = false;
        }

        public Triple(TreeNode root, boolean p, boolean q) {
            this.root = root;
            this.p = p;
            this.q = q;
        }

        public boolean found() {
            return p && q;
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Triple ans = this.f(root, p, q);
        return ans.root;
    }

    private Triple f(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return new Triple();
        }
        boolean isP = false;
        boolean isQ = false;
        if (root.equals(p)) {
            isP = true;
        }
        if (root.equals(q)) {
            isQ = true;
        }
        Triple left = this.f(root.left, p, q);
        if (left.found()) {
            return left;
        }
        Triple right = this.f(root.right, p, q);
        if (right.found()) {
            return right;
        }
        return new Triple(root, isP || left.p || right.p, isQ || left.q || right.q);
    }
}